### PR TITLE
Add meta data to task

### PR DESCRIPTION
**Pull request by kkoehn**
- add support for generic meta_data node on task level
  - make use of custom_namespace
  - use a nested hash in ruby as representation

### CODE REVIEW

**Review by MrSerth — COMMENTED** [review]
Great, thank you! 👏

File: lib/proforma/helpers/export_helpers.rb [hunk 1]
```diff
@@ -29,8 +29,34 @@ def add_test_configuration(xml, test)
...
+
+      def inner_meta_data(xml, namespace, data)
+        data.each do |key, value|
+          xml[namespace].send("#{key}_", value) if value.is_a? String
+
+          next unless value.is_a? Hash
+
+          xml[namespace].send("#{key}_") do |meta_data_xml|
+            inner_meta_data(meta_data_xml, namespace, value)
+          end
```
[the remark this answers is not in the document]
Comment (kkoehn) [reply lib/proforma/helpers/export_helpers.rb]: I'm still not sure about the Arrays. On the ruby-side it makes sense to use them, but in xml they would look like this.
That feels kind of awkward I think.
What do you think?
```
meta_data: {
  openHPI: {
    metao: ['first', 'second', 'third'],
  }
}
```
```
<meta-data>
  <openHPI:metao>first</openHPI:metao>
  <openHPI:metao>second</openHPI:metao>
  <openHPI:metao>third</openHPI:metao>
</meta-data>
```

[the remark this answers is not in the document]
Comment (MrSerth) [reply lib/proforma/helpers/export_helpers.rb]: Mh, I feel you. However, I think that the example you provided is the intended mechanism to include arrays in XML.

As we were unsure about the actual use case of arrays in XML and the fact that I would assume additional complexity for the import, we might also postpone the support for arrays here.

File: Gemfile.lock [hunk 2]
```diff
@@ -67,7 +67,7 @@ GEM
     pry-byebug (3.9.0)
       byebug (~> 11.0)
       pry (~> 0.13.0)
-    racc (1.6.0)
+    racc (1.5.2)
```
Comment (MrSerth) [inline Gemfile.lock:70]: Is downgrading `racc` an intended change?

File: bin/console [hunk 3]
```diff
@@ -24,6 +24,16 @@ Pry.start
 #   proglang: { name: 'Ruby', version: '1' },
 #   uuid: SecureRandom.uuid,
 #   language: 'de',
+#   meta_data: {
+#     CodeOcean: {
```
Comment (MrSerth) [inline bin/console:28]: Thanks for changing the example to use the `CodeOcean` namespace 🙏

File: lib/proforma/helpers/import_helpers.rb [hunk 4]
```diff
@@ -74,13 +74,17 @@ def test_files_from_test_configuration(test_configuration_node)
...
-
-          any_data_node.children.each do |any_data_tag|
-            any_data << {namespace: any_data_node.children.first.namespace.prefix, key: any_data_tag.name,
-                         value: any_data_tag.children.first.text}
+      def meta_data(any_data_node, use_namespace: false)
+        # use_namespace forces the use of the namespace as hash key - it should only be used at the entry of the recursion
+        {}.tap do |any_data|
+          any_data_node.children.each do |node|
+            key = (use_namespace ? node.namespace.prefix : any_data_node.name).to_sym
+            any_data[key.to_sym] = if node.node_type == Nokogiri::XML::Node::TEXT_NODE
```
Comment (MrSerth) [inline lib/proforma/helpers/import_helpers.rb:82]: Is it required to call the method `to_sym` here again? I would have expected that it is sufficient to have the call `to_sym` once in line 81?

File: lib/proforma/helpers/import_helpers.rb [hunk 5]
```diff
@@ -74,13 +74,17 @@ def test_files_from_test_configuration(test_configuration_node)
...
+                                     node.text
+                                   else
+                                     # preserve any existing data in the nested hash
+                                     (any_data[key.to_sym] || {}).merge meta_data(node)
```
Comment (MrSerth) [inline lib/proforma/helpers/import_helpers.rb:86]: Same remark with `key`

**Review by MrSerth — APPROVED** [review]
Awesome work, thank you! I have another minor question but think that we're almost ready to merge this! 🚀

File: Gemfile.lock (hunk 2 above)
Comment (kkoehn) [reply Gemfile.lock]: I don't know where did this come from - reverted!

File: lib/proforma/helpers/import_helpers.rb (hunk 4 above)
Comment (kkoehn) [reply lib/proforma/helpers/import_helpers.rb]: you are correct - fixed!

**Review by MrSerth — APPROVED** [review]
Thanks for tackling the few review points, I really like the changes and the new, clean API for importing and exporting metadata 👌

**Pull request merged**